test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java

```diff
@@ -63,34 +74,55 @@ public static void setup() throws InterruptedException, MarathonException, URISy
         //get the zk ip details and pass it to bk, host, controller
         URI zkUri = zkUris.get(0);
         //2, check if bk is running, otherwise start, get the zk ip
-        Service bkService = new BookkeeperService("bookkeeper", zkUri);
+        Service bkService = Utils.isDockerLocalExecEnabled() ?
+                new BookkeeperDockerService("bookkeeper", zkUri)
+                : new BookkeeperService("bookkeeper", zkUri);
         if (!bkService.isRunning()) {
             bkService.start(true);
         }
 
         List<URI> bkUris = bkService.getServiceDetails();
         log.debug("bookkeeper service details: {}", bkUris);
 
+        //start HDFS
+        URI hdfsUri = null;
+        if (Utils.isDockerLocalExecEnabled()) {
+            Service hdfsService = new HDFSDockerService("hdfs");
+            if (!hdfsService.isRunning()) {
+                hdfsService.start(true);
+            }
+            hdfsUri = hdfsService.getServiceDetails().get(0);
+            log.debug("HDFS service details: {}", hdfsService.getServiceDetails());
+        }
+
         //3. start controller
-        Service controllerService = new PravegaControllerService("controller", zkUri);
+        Service controllerService = Utils.isDockerLocalExecEnabled()
+                ? new PravegaControllerDockerService("controller", zkUri)
+                : new PravegaControllerService("controller", zkUri);
         if (!controllerService.isRunning()) {
             controllerService.start(true);
         }
-
-        //4. start test controller instances
-        Service testControllerService = new PravegaControllerService(TEST_CONTROLLER_SERVICE_NAME, zkUri);
-        if (!testControllerService.isRunning()) {
-            testControllerService.start(true);
-        }
-
+        controllerService.scaleService(2, true);
         List<URI> conUris = controllerService.getServiceDetails();
-        log.debug("Pravega Controller service instance details: {}", conUris);
-
-        List<URI> testConUris = testControllerService.getServiceDetails();
-        log.debug("Pravega test Controller service instance details: {}", testConUris);
+        log.info("conuris {} {}", conUris.get(0), conUris.get(1));
+        log.debug("Pravega Controller service  details: {}", conUris);
+        // Fetch all the RPC endpoints and construct the client URIs.
+        List<String> uris;
+        if (Utils.isDockerLocalExecEnabled()) {
```

Comment:
This class seems to be half if/else blocks. Perhaps the test could go into an abstract parent class and there could be two child classes for the two cases.

Reply:
This number reduced greatly with use of factory method. `createServiceInstance`